python/observation-publisher/publisher/observation_sink.py

```diff
@@ -0,0 +1,223 @@
+                model_version=self._model_version,
+            )
+        except ValidationFailure as e:
+            error_message = "\n".join([err.error_message() for err in e.errors])
+            print(f"Failed to log to Arize: {error_message}")
+            raise e
+        except Exception as e:
+            print(f"Failed to log to Arize: {e}")
+            raise e
+
+
+@dataclass_json
+@dataclass
+class BigQueryConfig:
+    project: str
+    dataset: str
+    ttl_days: int
+
+
+class BigQuerySink(ObservationSink):
+    def __init__(
+        self,
+        inference_schema: InferenceSchema,
+        model_id: str,
+        model_version: str,
+        project: str,
+        dataset: str,
+        ttl_days: int,
+    ):
+        super().__init__(inference_schema, model_id, model_version)
+        self._client = BigQueryClient()
+        self._inference_schema = inference_schema
+        self._model_id = model_id
+        self._model_version = model_version
+        self._project = project
+        self._dataset = dataset
+        table = Table(self.write_location, schema=self.schema_fields)
+        table.time_partitioning = TimePartitioning(type_=TimePartitioningType.DAY)
+        table.expires = datetime.now() + timedelta(days=ttl_days)
+        self._table: Table = self._client.create_table(exists_ok=True, table=table)
+
+    @property
+    def schema_fields(self) -> List[SchemaField]:
+        value_type_to_bq_type = {
+            ValueType.INT64: "INTEGER",
+            ValueType.FLOAT64: "FLOAT",
+            ValueType.BOOLEAN: "BOOLEAN",
+            ValueType.STRING: "STRING",
+        }
+
+        schema_fields = [
+            SchemaField(
+                name=self._inference_schema.prediction_id_column,
+                field_type="STRING",
+            ),
+            SchemaField(
+                name=PREDICTION_LOG_TIMESTAMP_COLUMN,
+                field_type="TIMESTAMP",
+            ),
+        ]
+        for feature, feature_type in self._inference_schema.feature_types.items():
+            schema_fields.append(
+                SchemaField(
+                    name=feature, field_type=value_type_to_bq_type[feature_type]
+                )
+            )
+        for (
+            prediction,
+            prediction_type,
+        ) in self._inference_schema.model_prediction_output.prediction_types().items():
+            schema_fields.append(
+                SchemaField(
+                    name=prediction, field_type=value_type_to_bq_type[prediction_type]
+                )
+            )
+
+        return schema_fields
+
+    @property
+    def write_location(self) -> str:
+        table_name = f"prediction_log_{self._model_id}_{self._model_version}".replace(
+            "-", "_"
+        ).replace(".", "_")
+        return f"{self._project}.{self._dataset}.{table_name}"
+
+    def write(self, dataframe: pd.DataFrame):
+        self._client.insert_rows_from_dataframe(dataframe=dataframe, table=self._table)
```

Comment:
is this using streaming insert API?

Reply:
Yes, it's streaming insert API.

Reply:
is it easy if we use [BigQuery Storage Write API](https://cloud.google.com/bigquery/docs/write-api#api_flow) instead ?